driver/src/price_finding/naive_solver.rs

```diff
@@ -111,113 +111,132 @@ impl NaiveSolver {
-                let executed_order = &mut executed_orders[i];
-                if order.sell_token == fee.token {
-                    let price_buy = prices[&order.buy_token];
-                    executed_order.sell_amount =
-                        executed_sell_amount(fee, executed_order.buy_amount, price_buy, BASE_PRICE);
-                } else {
-                    let price_sell = prices[&order.sell_token];
-                    executed_order.buy_amount = match executed_buy_amount(
-                        fee,
-                        executed_order.sell_amount,
-                        BASE_PRICE,
-                        price_sell,
-                    ) {
-                        Some(exec_buy_amt) => exec_buy_amt,
-                        None => return Ok(Solution::trivial()),
-                    };
-                }
-            }
+fn create_executed_orders(first_match: &Match, fee: &Option<Fee>) -> (ExecutedOrderPair, PriceMap) {
+    fn create_executed_order(order: &Order, sell_amount: u128, buy_amount: u128) -> ExecutedOrder {
+        ExecutedOrder {
+            account_id: order.account_id,
+            order_id: order.id,
+            buy_amount,
+            sell_amount,
         }
+    }
+
+    // Preprocess order to leave "space" for fee to be taken
+    let x = order_with_buffer_for_fee(&first_match.orders[0], fee);
+    let y = order_with_buffer_for_fee(&first_match.orders[1], fee);
 
-        let solution = Solution {
-            prices,
-            executed_orders: executed_orders
-                .map(|orders| orders.to_vec())
-                .unwrap_or_else(|| vec![]),
+    let create_orders = |x_sell_amount, x_buy_amount, y_sell_amount, y_buy_amount| {
+        [
+            create_executed_order(&x, x_sell_amount, x_buy_amount),
+            create_executed_order(&y, y_sell_amount, y_buy_amount),
+        ]
+    };
+
+    let mut prices = HashMap::new();
+    let executed_orders = match first_match.order_pair_type {
+        OrderPairType::LhsFullyFilled => {
+            prices.insert(x.buy_token, x.sell_amount);
+            prices.insert(y.buy_token, x.buy_amount);
+            create_orders(x.sell_amount, x.buy_amount, x.buy_amount, x.sell_amount)
+        }
+        OrderPairType::RhsFullyFilled => {
+            prices.insert(x.sell_token, y.sell_amount);
+            prices.insert(y.sell_token, y.buy_amount);
+            create_orders(y.buy_amount, y.sell_amount, y.sell_amount, y.buy_amount)
+        }
+        OrderPairType::BothFullyFilled => {
+            prices.insert(y.buy_token, y.sell_amount);
+            prices.insert(x.buy_token, x.sell_amount);
+            create_orders(x.sell_amount, y.sell_amount, y.sell_amount, x.sell_amount)
+        }
+    };
+
+    (executed_orders, prices)
+}
+
+/// Returns whether the solution should be trivial.
+fn update_with_fee(
+    orders: &OrderPair,
+    fee: &Fee,
+    executed_orders: &mut ExecutedOrderPair,
+    prices: &mut PriceMap,
+) -> bool {
```

Comment:
nit: I find this return type a bit un-intuitive. Should it maybe return an Option<()> so that `Some(())` means the update was successful? This would allow you to use a bunch of ? instead of match statements in this method. I wouldn't even mind if instead of mutating the orders and prices in place we returned an Option<(ExecutedOrderPair , PriceMap)> to make the funciton more pure.

Reply:
I made this take the arguments by value and return Solution. Then it is pure and can return `Solution::trivial()` itself.